File: lua/binary-search/binary-search.lua
return function (array, target)
  local function f(left, right)
    local len = right - left + 1
    local middle = left + math.floor(len / 2)

    if len <= 1 then return array[left] == target and left or -1 end

    local try = array[middle]

    if try < target then return f(middle + 1, right)
    elseif try > target then return f(left, middle - 1)
    else return middle end
  end

  return f(1, #array)
end

File: typescript/binary-search/binary-search.ts
export function find(haystack: number[], needle: number): number | never {
  let left = 0
  let right = haystack.length - 1
  let middle = 0

  while (left <= right) {
    middle = Math.floor((left + right) / 2)

    if (haystack[middle] > needle) {
      right = middle - 1
    }

    if (haystack[middle] < needle) {
      left = middle + 1
    }

    if (haystack[middle] == needle) {
      return middle
    }
  }

  throw Error("Value not in array")
}
